Core/src/main/java/org/tribuo/onnx/ONNXOperators.java

```diff
@@ -0,0 +1,196 @@
+     */
+    public final Map<String,ONNXAttribute> attributes;
+
+    /**
+     * Opset supported by these definitions.
+     */
+    private static final int OPSET_VERSION = 11;
+
+    private ONNXOperators(String value, int numInputs, int numOutputs) {
+        this.opName = value;
+        this.numInputs = numInputs;
+        this.numOutputs = numOutputs;
+        this.attributes = Collections.emptyMap();
+    }
+
+    private ONNXOperators(String value, int numInputs, int numOutputs, List<ONNXAttribute> attributes) {
+        this.opName = value;
+        this.numInputs = numInputs;
+        this.numOutputs = numOutputs;
+        Map<String,ONNXAttribute> attributeMap = new HashMap<>();
+        for (ONNXAttribute a : attributes) {
+            attributeMap.put(a.getName(),a);
+        }
+        if (attributes.size() != attributeMap.size()) {
+            throw new IllegalArgumentException("Duplicate attribute in enum declaration - " + attributes);
+        }
+        this.attributes = Collections.unmodifiableMap(attributeMap);
+    }
+
+    /**
+     * Builds this node based on the supplied inputs and outputs.
+     * Throws {@link IllegalArgumentException} if the number of inputs or outputs is wrong.
+     * @param context The onnx context used to ensure this node has a unique name.
+     * @param inputs The names of the inputs.
+     * @param outputs The names of the outputs.
+     * @return The NodeProto.
+     */
+    public OnnxMl.NodeProto build(ONNXContext context, String[] inputs, String[] outputs) {
+        return build(context,inputs,outputs,Collections.emptyMap());
+    }
+
+    /**
+     * Builds this node based on the supplied inputs and outputs.
+     * Throws {@link IllegalArgumentException} if the number of inputs, outputs or attributes is wrong.
+     * May throw {@link UnsupportedOperationException} if the attribute type is not supported.
+     * @param context The onnx context used to ensure this node has a unique name.
+     * @param inputs The names of the inputs.
+     * @param outputs The names of the outputs.
+     * @param attributeValues The attribute names and values.
+     * @return The NodeProto.
+     */
+    public OnnxMl.NodeProto build(ONNXContext context, String[] inputs, String[] outputs, Map<String,Object> attributeValues) {
+        if (inputs.length != numInputs) {
+            throw new IllegalArgumentException("Expected " + numInputs + " inputs, but received " + inputs.length);
+        }
+        if (outputs.length != numOutputs) {
+            throw new IllegalArgumentException("Expected " + numOutputs + " outputs, but received " + outputs.length);
+        }
+        if (attributeValues.size() > attributes.size()) {
+            throw new IllegalArgumentException("Found more attributes than expected, received " + attributeValues.size() + ", expected at most " + attributes.size());
+        }
+        if (!attributes.keySet().containsAll(attributeValues.keySet())) {
```

Comment:
Do we want to warn if `attributeValues` has an unrecognized key?

Reply:
Rather than throw an exception? My intention was that this check would throw if there was an unrecognized value.

I'd consider calling the build function with an unrecognized attribute to be a programming error which should be fixed, rather than something that we'd warn at runtime over.

Reply:
Sorry, I mean the other way around - if there are attributes defined by the operation that aren't provided by `attributeValues`. I guess some of them are required and others probably have implicit defaults, but there isn't a way to know that currently.

Reply:
I can extend the attribute type with if it's mandatory or not and then add that check. Whether an attribute is mandatory depends on the op set version number and Tribuo is going to pin to a specific version for a while.

Reply:
I think that might be worth doing, though I think you have a better sense than I of whether or not the added overhead is worth it in improved UX. My instinct upon reflection is that it might not be - most users won't be using these methods, and the error will end up getting thrown at runtime anyway.

Reply:
These classes are only for people extending Tribuo, not for people using it, so I'm not as worried about the amount of hand holding. I should properly document it, and it might be nice to mark the attributes so it's easy to see them in the enum, but I'm not sure if the code overhead is worthwhile.

Reply:
I documented and added the check, it wasn't too onerous.